docs/rfcs/0008-simplify-datastore.md

```diff
@@ -0,0 +1,126 @@
+
+# Detailed design
+
+As mentioned above, the current implementation of PipeCD allows us to just store objects as is, and from the reading side, we do not use complicated queries with `join` or something like that, which means we can simply write data objects to files directly. The remaining problems are:
+1. How to ensure the correctness to avoid data corruption, data loss, inconsistency from being handled by concurrent processes.
+2. How should we index/store the data objects so that we can read stored data fast enough to not make our users anger because of the slow response console.
+
+To handle (1) we have two approaches:
+- Achieve by implement __lock__
+- Achieve by avoid concurrent processes data as much as possible
+
+### Achieve by implement lock
+
+The most important thing to keep in mind while implement lock is to ensure there is always only ONE lock for each data object, and there will be only ONE process, which can hold the lock.
+With that note to keep in mind, we may have some approaches to go this way:
+- Create a single ambassador for write <1>
+- Using external service which controls locks (pessimistic locking) <2>
+- Using version field as data object properties and check for the version before write (optimistic locking) <3>
+
+#### <1> Create a single ambassador for write
+
+Use a single process to handling lock, we can achieve that by introducing a new PipeCD component which only handle data writing. This new component works as a proxy for the current datastore component, and we have to keep it always be singleton.
+
+![image](assets/single-ambassador.png)
+
+Cons:
+
+    - Downtime due to single point of failure.
+    - Performance on get since there is only one writer which handle all PipeCD components write requests.
+    - Have one more component which we need to handle ourself.
+
+#### <2> Using external service which controls locks (pessimistic locking)
+
+Use a external service or Cache (the current Cache component) to store lock for all objects. Whenever we need to write (create or update) objects, add one more step to ask the lock controlling service to get lock for the object.
+This approach is known as pessimistic locking.
+
+Pros:
+
+    - PipeCD's objects have unique id which can be used as the key for lock in Cache or external service easily.
+    - Already have a Cache component as a part of the control-plane, so we don't need to increase the number of components.
+
+Cons:
+
+    - Rely on an external system that holds the lock for all actors. Even in case we just use the current Cache component, it's designed to be faultable, it can be down but not effect the whole system. Making it controls locks means the Cache is must and become a high availability component -> increase the cost of operating.
```

Comment:
```suggestion
    - Rely on an external system that holds the lock for all actors.
      Even in case we just use the current Cache component, it's designed to be faultable, it can be down but not affect the whole system.
      Making it controls locks means the Cache is must and become a high availability component -> increase the cost of operating.
```